sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -674,13 +674,21 @@ public AverageRecordSize() {
       this.avgRecordGap = new MovingAvg();
     }
 
-    public void update(int recordSize, long gap) {
+    public synchronized void update(int recordSize, long gap) {
       avgRecordSize.update(recordSize);
       avgRecordGap.update(gap);
     }
 
     public double getTotalSize(double numRecords) {
-      return avgRecordSize.get() * numRecords / (1 + avgRecordGap.get());
+      double avgRecordSize;
+      double avgRecordGap;
+
+      synchronized (this) {
+        avgRecordSize = this.avgRecordSize.get();
+        avgRecordGap = this.avgRecordGap.get();
+      }
+
+      return avgRecordSize * numRecords / (1 + avgRecordGap);
```

Comment:
Not introduced by your change but why do we divide by the gap? why is it not just avgRecordSize * totalRecords?

Would be good to comment if there is a reason

Reply:
The next record offset for a partition is incremented for every record stored in that partition. Topic compaction only retains records with the greatest offset for the record's key within a partition, but the offsets remain untouched. The size of the backlog (`endOffset - beginningOffset`) may be greater than the number of records stored in the partition. Fetches from these partitions may return relatively few records compared to the expected size of the backlog. The average record gap scales the average record size to the expected fraction of retained records compared to the observed offsets. If I'm not mistaken, this estimator can be simplified to only use a (Guava) `AtomicDouble`.

Adding a `TODO` and some context in a comment block.